src/rust/aggregator/src/data/leaderboards.rs

```diff
@@ -0,0 +1,293 @@
+                FROM fill_events
+                WHERE time > $2 AND time < $3
+                AND NOT EXISTS (
+                    SELECT * FROM competition_indexed_events
+                    WHERE fill_events.txn_version = competition_indexed_events.txn_version
+                    AND fill_events.event_idx = competition_indexed_events.event_idx
+                    AND competition_indexed_events.competition_id = $1
+                )
+            )
+            INSERT INTO competition_leaderboard_users ("user", volume, integrators_used, n_trades ,competition_id)
+            SELECT "user", 0, '{}'::text[], 0, $1 FROM homogenous_fills WHERE "user" NOT IN (
+                SELECT "user" FROM competition_leaderboard_users WHERE competition_id = $1
+            )
+        "#,
+        comp.id,
+        comp.start,
+        comp.end,
+    )
+    .execute(transaction as &mut PgConnection)
+    .await
+    .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+    let users = sqlx::query!(
+        r#"
+            SELECT * FROM competition_leaderboard_users WHERE "user" IN (
+                SELECT "user" FROM fill_events
+                WHERE NOT EXISTS (
+                    SELECT * FROM competition_indexed_events
+                    WHERE fill_events.txn_version = competition_indexed_events.txn_version
+                    AND fill_events.event_idx = competition_indexed_events.event_idx
+                    AND competition_indexed_events.competition_id = $1
+                )
+                AND time > $2 AND time < $3
+            )
+            AND competition_id = $1
+        "#,
+        comp.id,
+        comp.start,
+        comp.end,
+    )
+    .fetch_all(transaction as &mut PgConnection)
+    .await
+    .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+    for user in users {
+        // TODO: dedup fills
+        // TODO: figure out why volume query can be null
+        sqlx::query!(
+            r#"
+                UPDATE competition_leaderboard_users
+                SET
+                    volume = volume + COALESCE((
+                        WITH relevant_fills AS (
+                            SELECT * FROM fill_events
+                            WHERE NOT EXISTS (
+                                SELECT * FROM competition_indexed_events
+                                WHERE fill_events.txn_version = competition_indexed_events.txn_version
+                                AND fill_events.event_idx = competition_indexed_events.event_idx
+                                AND competition_indexed_events.competition_id = $2
+                            )
+                            AND (maker_address = $1 OR taker_address = $1)
+                            AND time > $3 AND time < $4
+                        )
+                        SELECT SUM(size) FROM relevant_fills
+                    ), 0),
+                    integrators_used = COALESCE((
```

Comment:
Okay, I stopped reviewing this SQL about here. I know you worked very hard on it but please break it up and take advantage of Rust. You don't have to do it all at once in SQL.